docs: Revamp storage doc (#18259)
supabase/supabase: apps/docs/pages/guides/storage/uploads/resumable-uploads.mdx

    upload.stateFlow
    .onEach {
    println(it.progress)
  }
    .launchIn(yourCoroutineScope)

    upload.startOrResumeUploading()
  }

    //On other platforms you might have to give the bytes directly and specify a source if you want to continue it alter:
    suspend fun uploadData(bytes: ByteArray) {
    val upload: ResumableUpload = supabase.storage["bucket_name"].resumable.createOrContinueUpload(bytes, "source", "file_path")

    upload.stateFlow
    .onEach {
    println(it.progress)
  }
    .launchIn(yourCoroutineScope)

    upload.startOrResumeUploading()
  }
    ```

  </TabPanel>
</Tabs>

### Upload URL

When uploading using the resumable upload endpoint, the TUS server creates a unique URL for each upload, even for multiple uploads to the same path. All chunks will be uploaded to this URL using the `PATCH` method.

This URL will be valid for **up to 24 hours**. If the upload is not completed within 24 hours, the URL will expire and you'll need to start the upload again. The TUS client library will automatically create a new URL if the previous one expires.

### Concurrency

> When two or more clients try to upload to the same Upload URL only one of them will succeed. The other clients will receive a `409 Conflict` error. Only 1 client can upload to the same Upload URL at a time which prevents data corruption.

Reviewer: ```suggestion
When two or more clients try to upload to the same upload URL only one of them will succeed. The other clients will receive a `409 Conflict` error. Only 1 client can upload to the same upload URL at a time which prevents data corruption.
```

When two or more clients upload a file to the same path using different upload URLs, the first client to complete the upload will succeed and the other clients will receive a `409 Conflict` error.

If you provide the `x-upsert` header the last client to complete the upload will succeed instead.

### UppyJS Example

> You can check a full example using UppyJS [Here](https://github.com/supabase/supabase/tree/master/examples/storage/resumable-upload-uppy)

Reviewer: ```suggestion
You can check a [full example using UppyJS](https://github.com/supabase/supabase/tree/master/examples/storage/resumable-upload-uppy).
```

> Framework integration for UppyJS:

Reviewer: ```suggestion
Below are the different framework integrations for UppyJS:
```

- [React](https://uppy.io/docs/react/)
- [Svelte](https://uppy.io/docs/svelte/)
- [Vue](https://uppy.io/docs/vue/)
- [Angular](https://uppy.io/docs/angular/)

## Overwriting Files

When uploading a file to a path that already exists, the default behavior is to return a `400 Asset Already Exists` error.
If you want to overwrite a file on a specific path you can set the `x-upsert` header to `true`.

> We do advise against overwriting files when possible, as the CDN will take sometime to propagate the changes to all the edge nodes leading to stale content.

Reviewer: ```suggestion
We do advise against overwriting files when possible, as the CDN will take some time to propagate the changes to all the edge nodes leading to stale content.
```

Uploading a file to a new path is the recommended way to avoid propagation delays and stale content.

> If you want to know more about our CDN, check the [CDN](/docs/guides/storage/cdn) guide.

Reviewer: ```suggestion
To learn more, see the [CDN](/docs/guides/storage/cdn) guide.
```

## Access Control with RLS Policies

Both the standard file upload and TUS resumable file upload methods in Supabase Storage support Row-Level Security (RLS) policies. RLS allows you to define fine-grained access control rules to restrict access to files based on user roles, permissions, or other criteria. This ensures that your files are secure and accessible only to authorized users.

>   For more information on RLS policies, see the [Storage Access Control](/docs/guides/storage/access-control) guide.

Reviewer: ```suggestion
For more information on RLS policies, see the [Storage Access Control](/docs/guides/storage/access-control) guide.
```

## Conclusion

Supabase Storage provides multiple options for uploading files, including standard file upload and TUS resumable file upload, both of which use RLS policies for access control.

>   Additionally, there are client-side libraries like Uppy-js that make it easy to implement TUS uploads in your application. Choose the upload method that best fits your needs based on file size, network stability, and other requirements.

Reviewer: ```suggestion
Additionally, there are client-side libraries like Uppy-js that make it easy to implement TUS uploads in your application. Choose the upload method that best fits your needs based on file size, network stability, and other requirements.
```
